### DISCUSSION TITLE
Timer Wrapper

### DISCUSSION THREAD

**Discussion by beckydvn** [AI-Planning/macq#54] [General]
I'm wondering where the trace-timing wrapper should go file-wise. It's not pddl-centric in any way, it pretty much just takes a function and ensures it runs within the time given by the constant. Maybe in a separate file under the generate folder? Also, are there other wrappers we're going to need to eventually make?

**Comment by haz** [comment]
I think this one might be general enough. Maybe in `macq/utils`?

Note that this is kind of a special one -- we want to cap it across multiple function calls in order to capture "time since we've found something". There may be 2 attempts or 2000. So I can imagine usage kind of like...

```python
countdown = macq.utils.timer.Countdown(600) # 10min
traces = []
while len(traces) < num_traces:
    trace = compute_trace(...)
    if trace:
        traces.append(trace)
        countdown.reset()
    elif countdown.expired():
        # throw an error and explain that things timed out & how to change the constant
```

**Reply by beckydvn** [reply]
Hmm, so the timer is not a wrapper then? I guess generally a wrapper doesn't make sense for multiple function calls. (It works with my implementation because I put all the code to generate a single trace in a single function). That is, I have a function that loops until it finds a single trace, and that is what gets timed. However, I can see how that could cause problems in the future as we might not want to/be able to implement all trace generation that way.

Do you see any use in keeping my old wrapper around/refactoring it into something else, or should I delete it after making the `Countdown` class? Thanks for your input!

**Reply by haz** [reply]
Can you point to the old wrapper? I had something like [[this](https://stackoverflow.com/questions/2281850/timeout-function-if-it-takes-too-long-to-finish/22348885#22348885)] in mind when suggesting it. Problem is getting that "reset" type of functionality, but I guess if you have a `find_one_trace(...)` type of method that keeps trying until one's found, then _that_ could be wrapped...

**Reply by beckydvn** [reply]
Yep! Here's the wrapper:
https://github.com/QuMuLab/macq/blob/1fdad92f21d476f9fae09565b44e895548b2d5bd/macq/generate/pddl/generator.py#L17
And here is where it is used:
https://github.com/QuMuLab/macq/blob/1fdad92f21d476f9fae09565b44e895548b2d5bd/macq/generate/pddl/vanilla_sampling.py#L53

**Reply by haz** [reply]
Nice! So you went the decorator route. How UNIX-specific is that subprocess style?

Let's keep what you have, and...

1. rename to be more general (used in extraction and stuff too)
2. move to a top-level `utils` module
3. add a parameter for the decorator so we can pass in the # of seconds at the place of decoration

**Reply by beckydvn** [reply]
Will do! And according to the docs, it runs on both UNIX and Windows:
https://docs.python.org/3/library/multiprocessing.html

**Reply by beckydvn** [reply]
Just wondering, why add a parameter for the decorator if the user can change the MAX_TIME constant directly?

**Reply by haz** [reply]
Because we may want to define different constants and wrap multiple things with them. So the decorator uses the appropriate constant as its argument.
